mediatek: Add support for Acer Predator Connect W6x

ses1er opened the pull request:
Product name: Acer Predator Connect W6x
Product link: [https://www.acer.com/us-en/predator/networking/wi-fi/predator-connect-w6x/pdp/FF.G2TTA.001](https://www.acer.com/us-en/predator/networking/wi-fi/predator-connect-w6x/pdp/FF.G2TTA.001)

### Specifications:
SOC: MT7986AV
RAM: 1024MB
Flash: 256 MB SPI NAND
Ports: 4 LAN (1G) & 1 WAN (2.5G)
WIFI: MT7976GN + MT7976AN
LED: 1, ws2812b controller

### Installation via UART:
1. Configure TFTP server with IP 192.168.1.66. Copy `openwrt-mediatek-filogic-acer_predator-w6x-initramfs-kernel.bin` to TFTP root and rename to predator.bin
2. Interrupt boot by pressing 0 on startup or select `U-Boot Console` in U-Boot Boot Menu.
3. Run `setenv serverip 192.168.1.66; setenv ipaddr 192.168.1.1; tftpboot 0x46000000 predator.bin; fdt addr $(fdtcontroladdr); fdt rm /signature; bootm` in uboot console.
4. Wait for boot complete on Openwrt.
5. On client PC, transfer `openwrt-mediatek-filogic-acer_predator-w6x-squashfs-sysupgrade.bin` to `/tmp/` - `scp -O openwrt-mediatek-filogic-acer_predator-w6x-squashfs-sysupgrade.bin root@192.168.1.1:/tmp/sysupgrade.bin`
6. On router, run sysupgrade - `sysupgrade -n /tmp/sysupgrade.bin` 
7. Should now boot to Openwrt. Ensure it boots automatically to Openwrt by replugging the power.

Subsequent upgrades can be made by sysupgrade, or via Luci.

* LED color and intensity can be controlled by specifying RGB number in /sys/class/leds/rgb:status/multi_intensity Example: echo '75 0 130' > /sys/class/leds/rgb:status/multi_intensity

castiel652 commented:
Next time don't close the old PR.
You can just force push to same branch.

castiel652 commented on target/linux/mediatek/dts/mt7986a-acer-predator-w6x.dts:
```diff
@@ -0,0 +1,296 @@
+		reset-assert-us = <10000>;
+		reset-deassert-us = <10000>;
+		/* LED0: nc ; LED1: nc ; LED2: amber ; LED3: green */
+		mxl,led-config = <0x0 0x0 0x370 0x380>;
```
Change this to upstream style.
For reference https://github.com/openwrt/openwrt/commit/25ea7ff393a17d7bb3e037e9536143d9707e7d7a

ses1er replied:
thanks, changed.

castiel652 commented on target/linux/mediatek/dts/mt7986a-acer-predator-w6x.dts:
```diff
@@ -0,0 +1,311 @@
+};
+
+&slot0 {
+	radio0: mt7915@0,0 {
```
I don't think this is required since w6x doesn't have mt7916 for 6GHz.

ses1er replied:
you are correct, the w6x does not have 6Ghz.  Just tested and removing that block doesn't have any effect on the device.

ses1er replied:
i've also removed the associated kernel module (`kmod-mt7916-firmware`) as a dependency.

musashino205 commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -131,7 +131,8 @@ platform_do_upgrade() {
+	cudy,wr3000h-v1|\
+	acer,predator-w6x)
 		CI_UBIPART="ubi"
 		nand_do_upgrade "$1"
```
The default partition of `nand_do_upgrade` is `ubi` and this can be handled by the default section (`*)`) without device-specific entry.

https://github.com/openwrt/openwrt/blob/f9a8a3636c0079d01bcd2ea2458929c2f33a6458/package/base-files/files/lib/upgrade/nand.sh#L12

https://github.com/openwrt/openwrt/blob/f9a8a3636c0079d01bcd2ea2458929c2f33a6458/target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh#L212-L214

musashino205 commented on target/linux/mediatek/dts/mt7986a-acer-predator-w6x.dts:
```diff
@@ -0,0 +1,303 @@
+					};
+				};
+			};
+			factory: partition@380000 {
```
Please remove this unused DT label (`factory: `).

```suggestion
			partition@380000 {
```

ses1er replied:
👍

musashino205 commented on target/linux/mediatek/dts/mt7986a-acer-predator-w6x.dts:
```diff
@@ -0,0 +1,303 @@
+	pinctrl-0 = <&spi_flash_pins>;
+	status = "okay";
+
+	spi_nand_flash: spi_nand@0 {
```
ditto

```suggestion
	spi_nand@0 {
```

ses1er replied:
👍

musashino205 commented on target/linux/mediatek/dts/mt7986a-acer-predator-w6x.dts:
```diff
@@ -0,0 +1,303 @@
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+
+		partitions: partitions {
```
ditto

```suggestion
		partitions {
```

ses1er replied:
removed, thanks

abajk commented on target/linux/mediatek/dts/mt7986a-acer-predator-w6x.dts:
```diff
@@ -0,0 +1,303 @@
+		reset-assert-us = <10000>;
+		reset-deassert-us = <10000>;
+
+		leds {
```
Are you sure that the Ethernet port has built-in LEDs?

ses1er replied:
removed, thanks

ses1er commented on target/linux/mediatek/dts/mt7986a-acer-predator-w6x.dts:
```diff
@@ -0,0 +1,303 @@
+			>;
+		};
+	};
+};
```
i've reworked both files as suggested and tested to be working.

ses1er commented:
Thank you @goldwrt for your review and comments!

> To resolve the [“mislabeling issue” of the W6x](https://forum.openwrt.org/t/acer-predator-w6-with-openwrt/168979/256), you should use the same method that was used for “port@0“: The “label” is not defined in the common dts file but is defined in the device-specific dts files (“internet” for [W6m](https://github.com/openwrt/openwrt/blob/main/target/linux/mediatek/dts/mt7986a-acer-vero-w6m.dts), “game” for [W6](https://github.com/openwrt/openwrt/blob/main/target/linux/mediatek/dts/mt7986a-acer-predator-w6.dts) and [W6d](https://github.com/openwrt/openwrt/blob/main/target/linux/mediatek/dts/mt7986a-acer-predator-w6d.dts)).

I originally thought it was a labeling issue but port0 doesn't actually exist/work on the w6x, but rather moved by 1. (port0-port3 vs port1-port4 on the w6x). The labeling itself in this commit is consistent with the dtsi file.

> An alternative would be to include mt7986a.dtsi instead of mt7986a-acer-w6-common.dtsi in mt7986a-acer-predator-w6x.dts – because of the big hardware differences between w6x and W6/W6d/w6m (SPI NAND vs. eMMC, WS2812B vs. KTD2061/2026, ...).

I originally left the dtsi file alone and used `delete-node` to remove switch and refined the all the ports in the w6x dts file, but changed it due to suggestions by @musashino205 in previous comments. I understand the concerns of needing to modify the dtsi file and I agree the w6x has enough differences to warrant a new dtsi file but there are just as many similarities to other models.

I did go down the path to create a new dtsi file for the w6x, but reading the wiki/comments suggests to use a dtsi whenever possible. The result of result of using the dtsi file in this case resulted in a few `/delete-node/...` to make it work. I can certainly create a new dtsi for the w6x, but unsure what the consensus is on this.

ses1er commented:
@goldwrt @musashino205 

What do you both think about [this](https://github.com/openwrt/openwrt/compare/main...ses1er:predator-w6x-draft?expand=1)? This introduces minimal changes to the dtsi file, added only labels to minimize regression concerns. 

It builds correctly, but I need to pull my w6x from use to test it. If there's no concerns with this, i'll test, and merge the changes into this branch.

musashino205 commented:
> What do you both think about [this](https://github.com/openwrt/openwrt/commit/7496ddc4e6dae38ad6befa43d3cf6ac664ed0b78)?

LGTM.

BTW, placing pre-defined disabled ports is the same way as the mt7621.dtsi in the ramips target. In this way, all 5 ports on the internal MT7530 switch in the MT7621 SoC are difined with `status = "disabled";` and only necessary ports are enabled in dts of devices.
For W6* devices, since the MT7531 has 5 ports like the MT7621 and

- port@1-3 are commonly used on those devices ---> left enabled in the dtsi
- port@0 is used on most devices ---> left enabled in the dtsi and disabled on W6x
- port@4 is used only on W6x ---> added in disabled state to the dtsi

hauke commented:
Are the LAN port labels correct or are they also shifted by one? If they are shifted by one the dts file should update the label too.

ses1er commented:
> Are the LAN port labels correct or are they also shifted by one? If they are shifted by one the dts file should update the label too.

They're shifted by one. The labels are correct, eg, `lan1`matches port 1 on the device, etc.

----

After thinking about this further, I think the best approach is to`#include "mt7986a.dtsi"` rather than `#include "mt7986a-acer-w6-common.dtsi"`. This device has a lot of differences (as noted by @goldwrt in previous comments) to the Acer Predator/Vero w6* product line, even though they share similar product names. The deciding factor for me is this lays out the foundation for a potential uboot mod in the future, which I may try to tackle.

I've committed the standalone dts (`#include "mt7986a.dtsi"`) instead for review. I've tested it to build and run correctly.

ses1er committed:
mediatek: Add support for Acer Predator Connect W6x

Product name: Acer Predator Connect W6x
Product link: https://www.acer.com/us-en/predator/networking/wi-fi/predator-connect-w6x/pdp/FF.G2TTA.001

* Specifications:

SOC: MT7986AV
RAM: 1024MB
Flash: 256 MB SPI NAND
Ports: 4 LAN (1G) & 1 WAN (2.5G)
WIFI: MT7976GN + MT7976AN
LED: 1, ws2812b controller

* Installation via UART:

1. Configure TFTP server with IP 192.168.1.66. Copy `openwrt-mediatek-filogic-acer_predator-w6x-initramfs-kernel.bin` to TFTP root and rename to `predator.bin`
2. Interrupt boot by pressing 0 on startup or select `U-Boot Console` in U-Boot Boot Menu.
3. Run setenv `serverip 192.168.1.66; setenv ipaddr 192.168.1.1; tftpboot 0x46000000 predator.bin; fdt addr $(fdtcontroladdr); fdt rm /signature; bootm` in uboot console.
4. Wait for boot complete on Openwrt initramfs env.

** You can back up the MTD partitions at this point. Refer to Backup Instructions section.

5. On client PC, transfer `openwrt-mediatek-filogic-acer_predator-w6x-squashfs-sysupgrade.bin` to /tmp/ - `scp -O openwrt-mediatek-filogic-acer_predator-w6x-squashfs-sysupgrade.bin root@192.168.1.1:/tmp/sysupgrade.bin`
6. On router, run sysupgrade - `sysupgrade -n /tmp/sysupgrade.bin`
Should now boot to Openwrt. Ensure it boots automatically to Openwrt by replugging the power.

* Backup Instructions:

Layout from stock firmware:
```
[    0.968731] Creating 10 MTD partitions on "nmbm_spim_nand":
[    0.974297] 0x000000000000-0x000000100000 : "BL2"
[    0.979424] 0x000000100000-0x000000180000 : "u-boot-env"
[    0.985032] 0x000000180000-0x000000380000 : "Factory"
[    0.990379] 0x000000380000-0x000000580000 : "FIP"
[    0.995378] 0x000000580000-0x000000600000 : "prod"
[    1.000461] 0x000000600000-0x000000700000 : "dual"
[    1.005527] 0x000000700000-0x000000800000 : "pot"
[    1.010516] 0x000000800000-0x000006c00000 : "ubi"
[    1.015626] 0x000006c00000-0x00000d000000 : "ubi1"
[    1.020801] 0x00000d000000-0x00000d800000 : "storage"
```

Mapping in initramfs env:
```
dev:    size   erasesize  name
mtd0: 00100000 00020000 "bl2"
mtd1: 00080000 00020000 "u-boot-env"
mtd2: 00200000 00020000 "factory"
mtd3: 00200000 00020000 "fip"
mtd4: 00020000 00020000 "prod"
mtd5: 00100000 00020000 "dual"
mtd6: 00100000 00020000 "pot"
mtd7: 06400000 00020000 "ubi"
mtd8: 06400000 00020000 "ubi1"
mtd9: 00800000 00020000 "storage"
```
1. While in openwrt initramfs environment, back up all the partitions by running the following:
```
cat /dev/mtd0 > /tmp/bl2.bin
cat /dev/mtd1 > /tmp/u-boot-env.bin
cat /dev/mtd2 > /tmp/factgory.bin
cat /dev/mtd3 > /tmp/fip.bin
cat /dev/mtd4 > /tmp/prod.bin
cat /dev/mtd5 > /tmp/dual.bin
cat /dev/mtd6 > /tmp/pot.bin
cat /dev/mtd7 > /tmp/ubi.bin
cat /dev/mtd8 > /tmp/ubi1.bin
cat /dev/mtd9 > /tmp/storage.bin
```
2. Transfer files to client PC for safekeeping. On client PC, run `scp -O root@192.168.1.1:/tmp/*.bin ./`

* Restore to Stock Firmware:

1. Boot to openwrt initramfs env.
2. Confirm layout matches as follows by running `cat /proc/mtd`. Ensure dev `mtd7` is named `ubi`:
```
dev:    size   erasesize  name
mtd0: 00100000 00020000 "bl2"
mtd1: 00080000 00020000 "u-boot-env"
mtd2: 00200000 00020000 "factory"
mtd3: 00200000 00020000 "fip"
mtd4: 00020000 00020000 "prod"
mtd5: 00100000 00020000 "dual"
mtd6: 00100000 00020000 "pot"
mtd7: 06400000 00020000 "ubi"
mtd8: 06400000 00020000 "ubi1"
mtd9: 00800000 00020000 "storage"
```
3. Detach `ubi` partition - `ubidetach -p /dev/mtd7`
4. Transfer stock firmware's `ubi.bin` to router from client PC: `scp -O ubi.bin root@192.168.1.1/tmp/`
5. Format and replace with backup `ubiformat /dev/mtd7 -y -f /tmp/ubi.bin`
6. Reboot and you should now be back on stock firmware.

* LEDS:
LED color can be controlled by specifying values in GRB format in `/sys/class/leds/rgb:status/multi_intensity`. Default is `255 255 255` (white).
Example: `echo '75 0 130' > /sys/class/leds/rgb:status/multi_intensity`

LED brightness can be changed by specifying the value from 0-255 in /sys/class/leds/rgb:status/brightness. Default is `255` (full brightness).
Example: `echo 100 > /sys/class/leds/rgb:status/brightness`

For persistence across reboots, put the relevant command(s) in /etc/rc.local.

* Notes:

root access on stock firmware:
Before flashing openwrt, and while in openwrt initramfs env:
1. Mount /dev/ubi0_2: `mkdir /tmp/ubi0_2; mount -t ubifs /dev/ubi0_2 /tmp/ubi0_2`
2. Modify `/tmp/ubi0_2/upper/etc/passwd` and change line with `root:x:0:0...` to `root::0:0...`, remove the `x`.
3. Save and reboot.
4. You should now be able to log in with root and empty password while booted in stock firmware.

While on Openwrt, subsequent upgrades can be made by sysupgrade, or via Luci. UART should not be necessary unless you want to revert to stock firmeware.

Signed-off-by: Qing W <ses1er@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/19754
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
